src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,623 @@
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_validation_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+        best_epoch_info: Dict[str, Any],
+        ctx: mx.Context,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                    best_epoch_info=best_epoch_info,
+                    ctx=ctx,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_train_end(
+        self,
+        training_network: nn.HybridBlock,
+        temporary_dir: str,
+        ctx: mx.context.Context = None,
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_train_end(
+                training_network=training_network,
+                temporary_dir=temporary_dir,
+                ctx=ctx,
+            )
+
+
+class TrainingHistory(Callback):
+    @validated()
+    def __init__(self):
+        self.loss_history = []
+        self.validation_loss_history = []
+
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        self.loss_history.append(epoch_loss)
+        return True
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        self.validation_loss_history.append(epoch_loss)
+        return True
+
+
+class TerminateOnNaN(Callback):
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        if math.isnan(epoch_loss):
+            logging.warning(
+                f"TerminateOnNaN Callback initiated stop of training at epoch {epoch_no}."
+            )
+            return False
+        else:
```

Comment:
Here I generally like to not have an `else`, but instead do the return on the top-level.